### PR TITLE
Adding ability to edit database connection properties in VSCode extension

### PR DESCRIPTION
**Pull request by jhorvath**
Adding ability to open property sheet of database connection in VSCode extension.

### CODE REVIEW

**Review by dbalek — APPROVED** [review]
It's a pity that the `html4j` API was not used.

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (sdedic) [inline java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:49]: I've just tried a different approach: declare a string-valued enum-like with `Save`, 'Cancel` ... members, make an union `CommandType` type:
```
export const CommandKey = {
    Save : "Save",
    Error : "Error",
    Info : "Info",
    Cancel : "Cancel"
} as const;

export type CommandType = (typeof CommandKey)[keyof typeof CommandKey];

export type MessageCommon<T extends CommandType> = Typed<T>;
```
(rest of code unchanged) ... but I am not entirely sure about benefits of such approach, it does not save much.
But I'd at least suggest to make the union type should use an `CommandKey` enum's values, to save string duplication:

```
export enum CommandKey {
    Save = "Save",
    Error = "Error",
    Info = "Info",
    Cancel = "Cancel"
}

export type Command = CommandKey.Save | CommandKey.Cancel | CommandKey.Error | CommandKey.Info;
```

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (sdedic) [inline java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:57]: Why using type aliases and their unions is more desirable than using `interface`s and their inheritance ?

File: java/java.lsp.server/vscode/src/propertiesView/propertiesHtmlBuilder.ts
Comment (sdedic) [inline java/java.lsp.server/vscode/src/propertiesView/propertiesHtmlBuilder.ts:58]: We have these String literals used in `PropTypes`, `PropTypeMap` and here ... still there is not an enum/const that would give a compile-type symbol to these literals ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:83]: This shouldn't happen (dispatch to a handler that did not declare the command): error or a log ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:86]: Malformed command: shouldn't it throw or at least log ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:90]: Not sure if 'tree service unavailable' is a valid state of the system; if not, throw an error.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:115]: Should the client really receive the same response for for malformed set command (no 2nd arg) and for successful set (and if other `completedFuture(null)`s are not fixed, for node-not-found, tree-not-supported conditions as well) ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:110]: What is the semantic of calling set(node, null) ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:142]: This will make some entries of the returned array `null`, client having to check for null/undefined there. Wouldn't it be better to accumulate in a List and turn to array at the end ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:112]: Hm ... could be more friendly if the property sets was a map (keyed by set name) instead of array, but the propset name is present in each array's item, so it's probably OK.

Maybe if
```
List<Map<String, Object>> m = ...
```
the `setAllProperties` and others can use strongly typed Map instead of raw type(s) ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java:197]: Maybe there could be a way how to report a failed set to the client.

**Review by sdedic — CHANGES_REQUESTED** [review]
at least the unnecessary whitespace reformat should be fixed before merge.

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:14]: No, ATM...
Numbers aren't present in DB properties and would require special handling...

File: java/java.lsp.server/vscode/src/propertiesView/script.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/script.ts:16]: Sure, this is bad error message...

File: java/java.lsp.server/vscode/src/typesUtil.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/typesUtil.ts:37]: Sure, I'll make comment on this type declaration...

File: java/java.lsp.server/vscode/package.json
Comment (sdedic) [reply java/java.lsp.server/vscode/package.json:1042]: I was surprised; I understand that shell `|` operator starts both processes, but also connects (?) their stdins - for parallel runs I've seen techniques using `npm-run-all` or `concurrently` packages. But this is shorter as long as it works as expected.

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:49]: The code is made so it is imperative to be complete...
The Value of 'enum' has to be exactly the same as the Key and has to contain all of Type union and nothing more...
Found no other way of achieving this behavior...

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:30]: This way it is imperative that the value is of expected type, the compiler can then enforce correct typing...

File: java/java.lsp.server/vscode/src/propertiesView/script.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/script.ts:80]: It will throw an exception...

File: java/java.lsp.server/vscode/src/propertiesView/script.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/script.ts:68]: Fixed...

File: java/java.lsp.server/vscode/src/propertiesView/script.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/script.ts:58]: Fixed, to throw an error...

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:24]: This is taken from return type of NBLS command...

File: java/java.lsp.server/vscode/src/extension.ts
Comment (Ondrej-Douda) [reply java/java.lsp.server/vscode/src/extension.ts:24]: FIxed...

File: java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts
Comment (sdedic) [reply java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts:30]: It eventually could, but the typing bound is quite awkward and still requiresa manual reflection-like selection of the correct code path:
```
    switch (prop.propType) {
        case 'java.lang.String':
            out = makeStringAccess(prop);
            break;
        case 'java.lang.Boolean':
            out = makeBoolAccess(prop);
```
...  instead of 
```
function makeStringAccess(prop: Property<'java.lang.String'>) {
```
the function can as well as do
```
function makeStringAccess(prop: Property) {
```
and do `prop.propValue as String` .. without all the fancy, but in effect almost useless constructions.

**Review by sdedic — COMMENTED** [review]
OK, let's fix the 'exceptional' return values from nbls later - I've filed #6113 for that.

**Comment by mbien** [comment]
please don't merge PRs without squashing them properly. Not sure what happened here, e.g the merge commit is missing which makes it difficult to track commits back to their PR.

**Pull request merged**
